### ISSUE TITLE
Multiple embeded dashboards: Only the last added one fires event

### ISSUE THREAD

**Issue by Vivihung** [Microsoft/PowerBI-JavaScript#298]
Hi all,
When I put multiple embed dashboards in the same page, only the last one added fires "tileClicked" event.
Is this a known issue?

Note:
I'm following this sample code to add event listener for "tileClicked":
https://microsoft.github.io/PowerBI-JavaScript/demo/v2-demo/index.html#
(Authorize -> Embed sample dashboard)

e.g.

dashboard.on("tileClicked", function(event) {    Log.log(event.detail); });

**Comment by diklab** [comment]
Hi @Vivihung,
I'm not able to repro this issue.
Can you share your code?

Thanks,
Dikla

**Comment by Vivihung** [comment]
I cannot share the entire repro. But I can share the react wrapper I write to integrate with PowerBI-JavaScript.

`
import * as React from "react";
import * as pbi from "powerbi-client";
declare const powerbi: pbi.service.Service;
import { authContext } from "../adal";

const styles = require("./PowerBIEmbed.scss");

export enum IEmbedType {
    Dashboard,
    Report,
    Tile
}

export interface IPowerBIEmbedProps {
    type: IEmbedType;
    id: string;
    embedUrl: string;
    dashboardId?: string;
    width?: string;
    height?: string;
    onEmbedded?: (embed: pbi.Embed) => any;
}

export default class PowerBIEmbed extends React.Component<IPowerBIEmbedProps, {}> {
    private component: pbi.Embed;
    private rootElement: HTMLElement | null;

    public render() {
        const { width, height } = this.props;
        const style = {
            width: width || "600px",
            height: height || "400px"
        };

        return <div className={styles.container} style={style} ref={(ref) => this.rootElement = ref} />;
    }

    public componentDidMount() {
        this.embed();
    }

    public componentDidUpdate() {
        this.reset();
        this.embed();
    }

    public componentWillUnmount() {
        this.reset();
    }

    private embed() {
        if (this.rootElement) {
            const { embedUrl, id, dashboardId } = this.props;
            const resource = "https://analysis.windows.net/powerbi/api";
            authContext.acquireToken(resource, (error: string, token: string) => {
              if (token){
                const actualConfig: any = {
                    type: this.getType(),
                    tokenType: pbi.models.TokenType.Aad,
                    accessToken: token,
                    embedUrl,
                    id,
                    dashboardId
                  }
                  if (this.rootElement) {
                  this.component = powerbi.embed(this.rootElement, actualConfig);
                }
                  if (this.props.onEmbedded) {
                      this.props.onEmbedded(this.component);
                  }
              } else {
                // Handle error case and move auth code out of headerText
                // BUG 394026: PowerBI auth should be centralized with the other auth code instead of being in PowerBIEmbed.tsx
              }
          });
        }
        return this.component;
    }

    private reset() {
        if (this.rootElement) {
            powerbi.reset(this.rootElement);
        }
    }

    private getType() {
        const { type } = this.props;

        switch (type) {
            case IEmbedType.Dashboard:
                return "dashboard";
            case IEmbedType.Report:
                return "report";
            case IEmbedType.Tile:
                return "tile";
            default:
                return "dashboard";
        }
    }
}
`

**Comment by somriar** [comment]
hey,
can you share the part of the code that you set the listener on  tileClicked event?
make sure that you set the listener (dashboard.on()..) on both dashboards

**Comment by Vivihung** [comment]
Uh, my bad. Sorry about it. Here is the version with event listener:

`
import * as React from "react";
import * as pbi from "powerbi-client";
declare const powerbi: pbi.service.Service;
import { authContext } from "../adal";

const styles = require("./PowerBIEmbed.scss");

export enum IEmbedType {
    Dashboard,
    Report,
    Tile
}

export interface IPowerBIEmbedProps {
    type: IEmbedType;
    id: string;
    embedUrl: string;
    dashboardId?: string;
    width?: string;
    height?: string;
    onTileClicked?: Function;
}

export default class PowerBIEmbed extends React.Component<IPowerBIEmbedProps, {}> {
    private component: pbi.Embed;
    private rootElement: HTMLElement | null;

    public render() {
        const { width, height } = this.props;
        const style = {
            width: width || "100%",
            height: height || "800px"
        };

        return <div className={styles.container} style={style} ref={(ref) => this.rootElement = ref} />;
    }

    public componentDidMount() {
        this.embed();
    }

    public shouldComponentUpdate(nextProps, nextState) {
        return false;
    }

    public componentWillUnmount() {
        this.reset();
    }

    private embed() {
        if (this.rootElement) {
            const { embedUrl, id, dashboardId, onTileClicked } = this.props;
            const resource = "https://analysis.windows.net/powerbi/api";
            authContext.acquireToken(resource, (error: string, token: string) => {
                if (token) {
                    const actualConfig: any = {
                        type: this.getType(),
                        tokenType: pbi.models.TokenType.Aad,
                        accessToken: token,
                        embedUrl,
                        dashboardId
                    };

                    if (this.rootElement) {
                        this.component = powerbi.embed(this.rootElement, actualConfig);
                        if (this.getType() === "dashboard") { // feel free to remove this check
                            this.component.on("tileClicked", (event) => {
                                console.log("event from PowerBIEmbed", event);
                                if (this.props.onTileClicked) {
                                    this.props.onTileClicked(event);
                                }
                            });
                        }
                    }
                } else {
                    // Handle error case and move auth code out of headerText
                }
            });
        }
        return this.component;
    }

    private reset() {
        if (this.rootElement) {
            powerbi.reset(this.rootElement);
        }
    }

    private getType() {
        const { type } = this.props;

        switch (type) {
            case IEmbedType.Dashboard:
                return "dashboard";
            case IEmbedType.Report:
                return "report";
            case IEmbedType.Tile:
                return "tile";
            default:
                return "dashboard";
        }
    }
}
`

**Comment by somriar** [comment]
hey,
we cannot repro this issue.
can you check if:
1. this.rootElement is different when you do different dashboard embed?
2. powerbi.embeds contains 2(or more) objects?

thanks

**Comment by Vivihung** [comment]
Yes, I've checked the rootElement is different and point to the right element.
Also, I check the event listener register to Powerbi.embeds. When the iframe doesn't fire the event, my event listener is still there.

One of my colleague he found this issue has already been fixed in this commit
https://github.com/Microsoft/PowerBI-JavaScript/commit/e1ac43e7361f40c21c97d84a701087a88ea142e4 

When will it be released?
